### ISSUE TITLE
Tick lag when using 8+ Assemblers in a partially unloaded AE system

### ISSUE THREAD

**Issue by utoc** [Mordenkainen/EquivalentEnergistics#36]
This is an issue I brought up to you on the FTB subreddit.  I'll give you what I've got and lets see if its enough.

The trouble starts with chunks that are loaded by some buggy mod such that the entire AE network is not fully loaded.  The system in question had 14 EquivEnergistics assemblers connected to their network.  When the base was left for a user to go to other areas, the tick lag was immediately present, skipping 40-60 ticks until a restart when the chunks would be unloaded.

Most of the time the system would be idle, however there was some automated autocrafting present.  We removed the autocrafting and left the system idle and still the lag was present when the base was departed.

Ok, I have a sampler report: http://bbqsauce.terron.com/ee.nps

**Comment by utoc** [comment]
This may be unrelated but I am having a climbing entity count for "EMC Assembler".  When I first noticed it was at 1100, now we're up to 1347.  Is this related?

http://i.imgur.com/7wprpEU.png

There are only 19 EquivEnergistics crafting assemblers on the server.

**Comment by Mordenkainen** [comment]
WOAH!

I have no idea what would cause the TE's to keep climbing like that! That should not be happening!

OK, let me see if I can figure out why you appear to be leaking TE's.

**Comment by Mordenkainen** [comment]
So I was not able to recreate this. Though I did not recreate the "half loaded network" issue, I did cause chunks that contained several EMC Assemblers to load/unload several times and did not see huge increases in CPU usage or leaking TE's.

I suspect it may be cause by a mod that is either modifying the way chunk unloading happens, or by a block/mod that is doing chunkloading wrong.

I believe that the lag is caused because of the sheer number of leaked TE's. If a user has a single EMC Assembler, but it has leaked 100 TE's, then every time that user learns how o transmute a new item, it will have to inform 100 Assmblers of the change (instead of 1).

When I get he chance I will download your modpack and see if I can recreate the issue there.

**Comment by utoc** [comment]
What causes the TE to appear?  How do they leak to begin with?  I noticed after the server restart that the TE was cleared.  I may remove fastcraft just for grins one night to see if that has any effect.  I had never noticed the TE issue before, that was a first.  The partially loaded bases has been an issue though.  I will keep an eye on it and let you know if the TEs go hog wild again, it could just be a one-off thing.

**Comment by Mordenkainen** [comment]
Oh, I think they are directly related, based on the sampler data you provided.

Forgive me if I explain something you already know...

Blocks in Minecraft come in two "flavors". Blocks like cobble are static and have no logic, they can not be interacted with and dont interact with the game world. Blocks like a chest do have logic, and the logic is implemented by a special entity assiciated with the block called a TileEntity (TE for short).

Whenever Minecraft creates a block in the world, either because a user placed it or the chunk loads, the block is checked to see if it has a TE. If so one is created.

When the block is destroyed, or the chunk is unloaded, the TE is notified so that cleanup can be done.

The EMC Assembler has logic, so it has an associated TE. As part of the TEs creation, it adds itself to a list of EMC Assemblers in the world, so that events like EMC value changes can be sent to it. When the chunk is unloaded, or the block is destroyed, it removes itself from this list.

Due to how Java works, as long as the TE is in this list, Java will not allow it to be destroyed. This should not be a problem since when a TE is destroyed, it gets notified and should be removed from the list. It appears that in your case this notification is not happening, and the left over reference in the lost causes TEs to "leak", or hang around when they shouldn't. When the chunk hets loaded again, Minecraft doesn't know about the old TE, so creates a new one.

Because you have like 1300 extra EMC Assemblers hanging around, and those TEs have logic that must be processed 20 times a second, this causes the block lag.

I have some ideas I could try, if you would be willing to try running a special test build and rerunning the sampler to see if the problem still exists. That said I can only fix the side effects to my stuff. AE2 itself requires blocks to register and deregister from the network during these same events. If something in your pack is preventing these events from happening, you may find you will have far more serious problems over time.

**Comment by MaPePeR** [comment]
Maybe not store the TE itself, but a World Position (including dimension) in a Set and get the TE that way?

**Comment by Mordenkainen** [comment]
Actually, I think I will use a WeakReference, since that is what they are for anyway. An object reference that will not prevent garbage collection.

I'm still concerned that the TE's do not appear to be getting the invalidate() or onChunkUnload() calls though.

**Comment by MaPePeR** [comment]
they can't get `invalidate()` because they are still present in that List, i think. But why they don't get `onChunkUnload()` is a really mystery.

Didn't know about weak references in Java. So i learned something new today, thank you ;)

**Comment by Mordenkainen** [comment]
Yeah, they have been around since ~Java1.2, but still seem to be almost unknown.

**Comment by utoc** [comment]
Thank you for explaining TEs, I did not have any kind of grasp on that.  I will be glad to run a special build.. can you try your best to make it server-side only (so that I can run 0.5's release on my clients without them all needing to update to a special test build?)

**Comment by Mordenkainen** [comment]
Ok, try this build, you should only need to update the server. Monitor the situation and see if you are having the same issues.

https://dl.dropboxusercontent.com/u/53001205/EquivalentEnergistics-0.5-Test.jar

**Comment by utoc** [comment]
It'll be a couple of days before I can try this out, I'm about to push an update to the pack to get my users at 0.5 and then I'll slip the test build in and get back with you.

**Comment by Mordenkainen** [comment]
No problem, let me know how it goes.

**Comment by utoc** [comment]
I have an update.  0.5-Test does not seem to help the issue.  I have screenshots this time that may help.

First, profiling, I have two nps files:

http://bbqsauce.terron.com/ee-with-5assy.nps - This is a sampler report showing 5 EMC assemblers on an AE system that is not filly loaded.

http://bbqsauce.terron.com/ee-without.nps - This is a sampler report of the same dimension, all things the same except that the 5 assemblers and their tomes have been removed.  No other changes have been made.

This AE setup exists right next to a 5x5 area of loaded chunks.  This chunk is not force loaded, but instead because it has grass and grass blocks cause neighboring chunks to load and tick, it is sometimes loaded.

Without EMC crafters:  https://i.imgur.com/KgKUFDr.png

With EMC Crafters: https://i.imgur.com/7SI0QCh.png

I only let the samplers run for 70 seconds or so and we got down to 10tps.  Yesterday we were down to skipping 2600+ticks before I logged in and found the issue.  Force loading the chunks solves the problem and its what we're doing in the meantime.

I hope I've passed on more meaningful information.  Thank you for taking the time to see to all of this, you're awesome!

**Comment by Mordenkainen** [comment]
Did you check opis to see if you have leaking TEs?

**Comment by utoc** [comment]
I'll leave it running for six hours and report back.

I had a user report a possibly unrelated issue.  He was trying to make 100,000 cakes with EquivEnergistics and left his base while it was crafting.  He came back to 4000 EMC crystals all around his EMC assemblers.  Related?  Do you want me to open up a new issue on this?  I asked him to post on the FTB forum so you could see his screenshots.

**Comment by utoc** [comment]
Six hours later and we're still sitting at 17 like we were when we restarted.

**Comment by Mordenkainen** [comment]
That particular issue is an issue with AE.

If part of your crafting CPU unloads, it breaks the multiblock, causing it's contents to be dropped.

**Comment by Mordenkainen** [comment]
Was that 6 hours without the chunkloader?

**Comment by Mordenkainen** [comment]
OK, finally got a chance to look at your sampler reports.

I need to improve the efficiency of much of the code. It looks like there is a ton of overhead associated with building the patterns and reading the available transmutations. I will try to work out some caching to try to cut that down as much as possible.

I don't see evidence of a TE leak anymore.

In the short term, just keep the chunks loaded. Longer term I will try optimizing the code. Also, some of the overhead will probably change when I stop using AE internal code I shouldn't.

**Comment by utoc** [comment]
No, six hours with a chunkloader.  It is not possible to keep the server running without one.  We went from 20tps down to 10tps in about 30 seconds.  I have told my users to make sure that chunks are forceloaded in the meantime if they use the EMC assembler.

**Comment by Mordenkainen** [comment]
Ok, this is gonna take a bit for me to engineer some kind of recreation.

**Comment by Mordenkainen** [comment]
OK, some further info, most of this will probably be gibberish to you, but it still is good info.

Looking at the sampler output for the 5 Assemblers I can see that AE called the getPatterns() method ~ 5000 times. This is a slow and costly method, so I have it set up to only run when needed. The events that cause it to be run are listed above. 

All of these events should throw an MENetworkCraftingPatternChange Event, except for chunk loading and AE network reboot. The sampler shows 0 of this event being sent, so it would appear that the patterns are being re-read because of a network reboot or chunk load.

Going back to the previous math, this means that a chunk reloaded or the AE network rebooted ~1000 times in 70 seconds, or about every 1.5 ticks!

The sampler report without the EMC Assemblers shows the same thing happening with the AE Interfaces, but they are much less "costly" when reading their patterns so you probable don't notice the performance hit, but it is there. So it would appear this out of control re-reading the patterns is either due to some bug on your server, or the way the network is built.

This should NEVER happen!!!!!!

Again, I will do what I can to reduce the "cost" or reading the transmutation patterns, which should help... But won't solve the core issue.

TLDR: The problem with the Assemblers is a symptom of a much larger issue with either your server or the structure of your AE network.

**Comment by MaPePeR** [comment]
Is this with ProjectE or with EE3?

Do we need to do something on our side to make the Knowledge better? (At the moment that's a linear search on our side, so that is probably not a good thing. Is that causing parts of your issue?)

**Comment by Mordenkainen** [comment]
This is with EE3.

**Comment by utoc** [comment]
Another FTB power user and I discussed tick lag at length and we've come up with a few things.  One is that grass blocks on chunk borders will make neighboring chunks tick.  This is a Minecraft issue because of how grass spreads and this causes chunks to load and unload often throughout a day.  The AE system I showed you resides in two chunks and both border a 5x5 section that is constantly loaded.  This chunk also has grass.  Whenever force loaded chunks have grass ticks, the two chunks containing AE also tick.

I think one good test might be to put the crafting processor in the same chunk as the rest of the AE system and see if the symptoms are as severe.  I agree with you that the EMC assembler's tick lag is the symptom of a larger problem and that it isn't at fault so much as an innocent bystander.  To further prove the point, would I be correct in assuming that an EE tome that contained very little in the way of recipes shouldn't have such an exaggerated tick lag?  Would that be a worthwile test for you?  Now that I have a good test scenario, I can duplicate and diagnose this issue easily for testing.

**Comment by Mordenkainen** [comment]
That could be a good test.

Also, while I understand the odd behaviors grass can produce, it still doesnt explain your samples showing ~1000 pattern relearns over 70 seconds.

**Comment by Mordenkainen** [comment]
utoc,

I have another test version that I hope will resolve the majority of issues you have been seeing.

Would it be possible for you to test this, with no chunkloaders, and see if you get the same lag issues as before?

**Comment by utoc** [comment]
Any time, point me to the test version when you have it ready.

**Comment by Mordenkainen** [comment]
Thanks!

Please try this build.

It should work only updating the server, but if you continue to see the same issues, please try an updated client as well if possible.

I have largely rewritten how the patterns are handled from scratch. There is a lot more caching going on, so it should not try to recalculate all the patterns every time AE asks.

https://dl.dropboxusercontent.com/u/53001205/EquivalentEnergistics-0.5-Test2.jar

**Comment by utoc** [comment]
NPE Crash on startup with Test2: http://pastebin.com/8VfGwbyj

**Comment by Mordenkainen** [comment]
Ok, I know what the issue is. I'll have another build avail tmorrow morning.

**Comment by Mordenkainen** [comment]
@utoc Try this one:
https://dl.dropboxusercontent.com/u/53001205/EquivalentEnergistics-0.5-Test3.jar

**Comment by utoc** [comment]
NPE with 0.5-Test3 while loading world on a dedicated server: http://pastebin.com/TGTGeARP

Relevant forge log lines: http://pastebin.com/GZbKE9rj

**Comment by Mordenkainen** [comment]
Well now, that is interesting, and somewhat concerning.

That basically tells me that AE came up so fast and started calling methods EE3 before EE3 fonished calculating its EMC values.

**Comment by Mordenkainen** [comment]
Ok, think I found the issue. Try this:
https://dl.dropboxusercontent.com/u/53001205/EquivalentEnergistics-0.5-Test4.jar

**Comment by utoc** [comment]
Server loaded properly this time.  Unfortunately when we removed the chunkloader at the base in question, the lag returned as before.  I've taken a few sampler reports for you below.

Present at the base, 5 EMC Assemblers: http://bbqsauce.terron.com/ee-present.nps
Absent from his base, AE loaded but no crafting: http://bbqsauce.terron.com/ee-absent.nps

These were both 60 second samples, when we left his base (without a chunkloader going) we got down to 13.2 TPS per sfPlayer's Sampler mod.

I did not see an exaggerated tile entity count.

**Comment by Mordenkainen** [comment]
Thanks!

While I know it might not feel like we are making progress, we are.

The profiler outputs show that the leaking TE issue was indeed fixed, which improved the situation a lot.

In addition, the profiler shows that overall Equvalent Energistics is using 1/10th - 1/60th the CPU time after the changes, a significant improvement.

Doing this has removed a lot of "noise" from the profiler output and has allowed me to identify a few more places where the code can be optimized. It has also allowed me to identify exactly what is causing these "expensive" methods to be called over and over again, and why I was not seeing any of the trigger events happening.

I have some further optimizations to do, and will have another build for you to try.

Thanks for being patient while I puzzled my way through this.

**Comment by utoc** [comment]
Great news! We will await the next test version.

**Comment by Mordenkainen** [comment]
Please try this version:
https://dl.dropboxusercontent.com/u/53001205/EquivalentEnergistics-0.5-Test5.jar

Even if it appears to have no problem, I would still like a sampler report with no one near the chunk.

This situation seems to really stress the mod, and it would be nice to see if there are any more parts that should be optimized.

**Comment by utoc** [comment]
The Test5 version has a new item so I did a sample report from the console.  We still had considerable lag but since I didn't login, I don't know if improvements were made. If you'd like me to do some checking with Opis, let me know and I'll update my client as well.  Sample report is here: http://bbqsauce.terron.com/test5.nps and it is a 60 second sample.

Config files are located here: https://github.com/utoc/bbqClient/tree/master/minecraft
Give me an email address or some other way to message you to give you the URL to the world download zip file.

A good working example is in rfTools dimension 9.  You can get there with the command: `/rftdim tp 9 12 2 58` He has an AE system with 5 EMC assemblers.  You can place a ChickenChunks spot loader in the chunk located at X/Z -11, 66 to cause the symptoms to start.. It'll look something like this: http://i.imgur.com/reNwSpE.png

Thanks again for the help, looking forward to the next test

**Comment by Mordenkainen** [comment]
I sent you a PM on the FTB forums.

**Comment by Mordenkainen** [comment]
Ok, now that I got a look at the NPS file, the situation seems to be MUCH better. If there are lag issues in Equivalent Exchange, your profiler output does not show it.

Yay!

I still want to steal your world files though as it seems like it would be an excellent test environment.

**Comment by Mordenkainen** [comment]
Closing as I believe this issue was resolved.
